runtime/src/iree-amd-aie/aie_runtime/iree_aie_runtime.h

```diff
@@ -68,6 +68,33 @@ enum class AMDAIETileType : uint8_t {
   MAX = 4
 };
 
+/// Enum of DMA properties, using the offset within the `XAie_DmaMod` structure
+/// as underlying value.
```

Comment:
Can you motivate why this isn't just 0,1,2,3,... in the comment ? This design (I've not seen it before) makes me slightly nervous about layouts changing depending on the compiler. That said, I do like how it ties the underlying class in. What is MAX useful for? These are definitely not static members of XAie_DmaMod (seems like that'd be UB https://en.cppreference.com/w/cpp/named_req/StandardLayoutType)

Reply:
> What is MAX useful for?

:(

Reply:
lol not THAT max

Reply:
> Can you motivate why this isn't just 0,1,2,3,... in the comment ? This design (I've not seen it before) makes me slightly nervous about layouts changing depending on the compiler.

Could you elaborate on what you mean with layouts changing depending on the compiler? The idea of using `offsetof` is exactly to accommodate the XAie struct changing. 

> What is MAX useful for?

It can be useful to perform checks.

Reply:
I'm not sure if the compiler can rearrange the members to improve alignment. For example in

```
struct X{
 char a; 
 float b;
 char c;
uint16_t d;
 };
 ```
 
 Can the compiler effectively rearrange this so that it is 
 
 ```
 struct X {
 float b;
 char a; 
 char c;
 uint16_t d;
  };
 
 So that b is not offset by 1 byte.

Reply:
I see, this is not allowed I believe. Would introduce many issues if it were.

Reply:
Ok I did a bit of reading and indeed it's not allowed. Padding is allowed, but no reording.

Reply:
Yeah, from the `offsetof` doc:

> The macro offsetof expands to an integral constant expression of type [std::size_t](https://en.cppreference.com/w/cpp/types/size_t), the value of which is the offset, in bytes, from the beginning of an object of specified type to its specified subobject, including [padding bits](https://en.cppreference.com/w/cpp/language/object#Object_representation_and_value_representation) if any.

> Given an object o of type type and static storage duration, o.member shall be an lvalue constant expression that refers to a subobject of o. Otherwise, the behavior is undefined. Particularly, if member is a [static data member](https://en.cppreference.com/w/cpp/language/static), a [bit-field](https://en.cppreference.com/w/cpp/language/bit_field), or a [member function](https://en.cppreference.com/w/cpp/language/member_functions), the behavior is undefined.

>If type is not a [PODType](https://en.cppreference.com/w/cpp/named_req/PODType)(until C++11)[standard-layout](https://en.cppreference.com/w/cpp/language/data_members#Standard-layout) type(since C++11), the result of offsetof is undefined(until C++17)use of the offsetof macro is conditionally-supported(since C++17).

>
> The expression offsetof(type, member) is never [type-dependent](https://en.cppreference.com/w/cpp/language/dependent_name#Dependent_types) and it is value-dependent if and only if type is dependent.